### PR TITLE
feat(account-lib): rebase account-lib in the bitgoJs and fixing errors

### PR DESCRIPTION
**Pull request by JoaquinTissera-Simtlix**
rebase account-lib in the bitgoJs and fixing errors

STLX-5609

### CODE REVIEW

File: modules/account-lib/src/coin/algo/utils.ts [hunk 1]
```diff
@@ -29,6 +31,26 @@ function allHexChars(maybe: string): boolean {
...
 
+function arrayEqual(a, b) {
+  if (a.length !== b.length) {
+    return false;
+  }
+  return a.every((val, i) => val === b[i]);
+}
+
+/**
+ * ConcatArrays takes two array and returns a joint array of both
```
Comment (rliberal-simtlix) [inline modules/account-lib/src/coin/algo/utils.ts:42]: Seems to be more generic util. Please check if there a more generic utils file.

File: modules/account-lib/src/coin/algo/utils.ts [hunk 2]
```diff
@@ -64,6 +86,10 @@ export class Utils implements BaseUtils {
     throw new NotImplementedError('hash not implemented.');
   }
 
+  areKeysEqual(key1: Uint8Array, key2: Uint8Array): boolean {
```
Comment (rliberal-simtlix) [inline modules/account-lib/src/coin/algo/utils.ts:89]: Add js doc

File: modules/account-lib/src/coin/baseCoin/ed25519KeyPair.ts [hunk 3]
```diff
@@ -99,9 +103,24 @@ export abstract class Ed25519KeyPair implements BaseKeyPair {
...
    */
-  verifySignature(message: string, signature: Uint8Array): boolean {
-    const messageToVerify = toUint8Array(Buffer.from(message).toString('hex'));
+  verifySignature(message: Uint8Array | string, signature: Uint8Array): boolean {
+    const messageToVerify = this.toBuffer(message);
     const publicKey = toUint8Array(this.keyPair.pub);
     return nacl.sign.detached.verify(messageToVerify, signature, publicKey);
   }
+
+  toBuffer(message: string | Buffer | Uint8Array) {
```
Comment (rliberal-simtlix) [inline modules/account-lib/src/coin/baseCoin/ed25519KeyPair.ts:112]: add js doc

File: modules/account-lib/src/coin/algo/utils.ts [hunk 4]
```diff
@@ -29,6 +31,26 @@ function allHexChars(maybe: string): boolean {
   return maybe.match(/^[0-9a-f]+$/i) !== null;
 }
 
+function arrayEqual(a, b) {
```
Comment (rliberal-simtlix) [inline modules/account-lib/src/coin/algo/utils.ts:34]: Add js doc

Comment (rliberal-simtlix) [reply modules/account-lib/src/coin/algo/utils.ts:34]: Try to use this. https://www.geeksforgeeks.org/lodash-_-isequal-method/

File: modules/account-lib/src/coin/algo/utils.ts (hunk 1 above)
Comment (rliberal-simtlix) [reply modules/account-lib/src/coin/algo/utils.ts:42]: https://dustinpfister.github.io/2018/08/02/lodash_concat/

**Review by rliberal-simtlix — CHANGES_REQUESTED** [review]

File: modules/account-lib/src/coin/algo/keyPair.ts [hunk 5]
```diff
@@ -17,12 +19,13 @@ export class KeyPair extends Ed25519KeyPair {
 
   /** @inheritdoc */
   recordKeysFromPrivateKeyInProtocolFormat(prv: string): DefaultKeys {
-    throw new NotImplementedError('recordKeysFromPrivateKeyInProtocolFormat not implemented');
+    throw new NotImplementedError('recordKeysFromPublicKeyInProtocolFormat not implemented');
```
Comment (JulioAchaval-SimTLiX) [inline modules/account-lib/src/coin/algo/keyPair.ts:22]: Is this error ok?

File: modules/account-lib/src/coin/algo/utils.ts (hunk 4 above)
Comment (JulioAchaval-SimTLiX) [reply modules/account-lib/src/coin/algo/utils.ts:34]: +1

File: modules/account-lib/src/coin/algo/utils.ts (hunk 1 above)
Comment (JulioAchaval-SimTLiX) [reply modules/account-lib/src/coin/algo/utils.ts:42]: +1

File: modules/account-lib/src/coin/algo/utils.ts [hunk 6]
```diff
@@ -89,48 +115,82 @@ export class Utils implements BaseUtils {
    * @param {string} seed - the seed to be validated
    * @returns {boolean} - true if the seed is valid
    */
-   isValidSeed(seed: string): boolean {
-    if (typeof seed !== "string") return false;
-    if (seed.length !== ALGORAND_SEED_LENGTH) return false; // Try to decode
+  isValidSeed(seed) {
```
Comment (JulioAchaval-SimTLiX) [inline modules/account-lib/src/coin/algo/utils.ts:118]: Add types of param and return

File: modules/account-lib/src/utils/crypto.ts [hunk 7]
```diff
@@ -144,7 +144,7 @@ export function isValidEd25519SecretKey(prv: string): boolean {
  */
 export function isValidEd25519PublicKey(pub: string): boolean {
   try {
-    const decodedPub = toUint8Array(pub);
+    const decodedPub = new Uint8Array(Buffer.from(pub, 'hex')); // toUint8Array(pub);
```
Comment (JulioAchaval-SimTLiX) [inline modules/account-lib/src/utils/crypto.ts:147]: What is the purpose of this comment?

File: modules/account-lib/test/unit/coin/algo/transactionBuilder/assetTransferBuilder.ts [hunk 8]
```diff
@@ -64,15 +64,11 @@ describe('Algo Asset Transfer Transaction Builder', () => {
...
-    it('should build a valid disable token transaction', async function() {
+    it('should build a valid disable token transaction', async function () {
       const firstRound = 167;
       const lastRound = 1167;
       const fee = 1000;
       const tokenId = 1;
-      const {
-        networks: { testnet },
-      } = AlgoResources;
-      const { genesisHash, genesisID } = testnet;
```
Comment (JulioAchaval-SimTLiX) [inline modules/account-lib/test/unit/coin/algo/transactionBuilder/assetTransferBuilder.ts:75]: Why you delete this lines?

File: modules/account-lib/test/unit/coin/algo/transactionBuilder/assetTransferBuilder.ts [hunk 9]
```diff
@@ -110,33 +106,29 @@ describe('Algo Asset Transfer Transaction Builder', () => {
...
+    it('should build a valid enable token transaction', async function () {
       const firstRound = 167;
       const lastRound = 1167;
       const fee = 1000;
       const tokenId = 1;
-      const {
-        networks: { testnet },
-      } = AlgoResources;
-      const { genesisHash, genesisID } = testnet;
       const amount = '0';
```
Comment (JulioAchaval-SimTLiX) [inline modules/account-lib/test/unit/coin/algo/transactionBuilder/assetTransferBuilder.ts:132]: Same here

File: modules/account-lib/test/unit/coin/algo/transactionBuilder/assetTransferBuilder.ts [hunk 10]
```diff
@@ -168,15 +160,11 @@ describe('Algo Asset Transfer Transaction Builder', () => {
...
-    it('should build a valid enable token transaction and re-calcualte minimum fee', async function() {
+    it('should build a valid enable token transaction and re-calcualte minimum fee', async function () {
       const firstRound = 167;
       const lastRound = 1167;
       const fee = 1000;
       const tokenId = 1;
-      const {
-        networks: { testnet },
-      } = AlgoResources;
-      const { genesisHash, genesisID } = testnet;
```
Comment (JulioAchaval-SimTLiX) [inline modules/account-lib/test/unit/coin/algo/transactionBuilder/assetTransferBuilder.ts:179]: Same here

File: modules/account-lib/test/unit/coin/algo/transactionBuilder/assetTransferBuilder.ts [hunk 11]
```diff
@@ -208,16 +196,12 @@ describe('Algo Asset Transfer Transaction Builder', () => {
...
+    it('should build a valid disable token transaction and re-calcualte minimum fee', async function () {
       const firstRound = 167;
       const lastRound = 1167;
       const fee = 1000;
       const tokenId = 1;
       const closeRemainderTo = 'SP745JJR4KPRQEXJZHVIEN736LYTL2T2DFMG3OIIFJBV66K73PHNMDCZVM';
-      const {
-        networks: { testnet },
-      } = AlgoResources;
-      const { genesisHash, genesisID } = testnet;
```
Comment (JulioAchaval-SimTLiX) [inline modules/account-lib/test/unit/coin/algo/transactionBuilder/assetTransferBuilder.ts:220]: Same here

**Review by JulioAchaval-SimTLiX — CHANGES_REQUESTED** [review]

File: modules/account-lib/test/unit/coin/algo/transactionBuilder/assetTransferBuilder.ts (hunk 8 above)
Comment (Krin2) [reply modules/account-lib/test/unit/coin/algo/transactionBuilder/assetTransferBuilder.ts:75]: This lines weren't used and they be marked by lint as an error

File: modules/account-lib/test/unit/coin/algo/transactionBuilder/assetTransferBuilder.ts (hunk 9 above)
Comment (Krin2) [reply modules/account-lib/test/unit/coin/algo/transactionBuilder/assetTransferBuilder.ts:132]: This lines weren't used and they be marked by lint as an error

File: modules/account-lib/test/unit/coin/algo/transactionBuilder/assetTransferBuilder.ts (hunk 10 above)
Comment (Krin2) [reply modules/account-lib/test/unit/coin/algo/transactionBuilder/assetTransferBuilder.ts:179]: This lines weren't used and they be marked by lint as an error

File: modules/account-lib/test/unit/coin/algo/transactionBuilder/assetTransferBuilder.ts (hunk 11 above)
Comment (Krin2) [reply modules/account-lib/test/unit/coin/algo/transactionBuilder/assetTransferBuilder.ts:220]: This lines weren't used and they be marked by lint as an error

File: modules/account-lib/test/unit/coin/algo/transactionBuilder/assetTransferBuilder.ts (hunk 8 above)
Comment (JulioAchaval-SimTLiX) [reply modules/account-lib/test/unit/coin/algo/transactionBuilder/assetTransferBuilder.ts:75]: You're right. You can use builder.testnet(); instead this approach.

**Review by JulioAchaval-SimTLiX — APPROVED** [review]
LGTM

File: modules/account-lib/src/coin/algo/keyPair.ts [hunk 12]
```diff
@@ -1,8 +1,10 @@
 /* eslint-disable no-unused-vars */
-import { encodeAddress } from 'algosdk';
+import * as nacl from 'tweetnacl';
+import algosdk, { encodeAddress } from 'algosdk';
 import { Ed25519KeyPair } from '../baseCoin';
 import { NotImplementedError } from '../baseCoin/errors';
 import { DefaultKeys, KeyPairOptions } from '../baseCoin/iface';
+import { toHex, toUint8Array } from '../hbar/utils';
```
Comment (rliberal-simtlix) [inline modules/account-lib/src/coin/algo/keyPair.ts:7]: why hbar utils? There no common utils for this?

File: modules/account-lib/src/coin/algo/transactionBuilder.ts [hunk 13]
```diff
@@ -292,8 +292,14 @@ export abstract class TransactionBuilder extends BaseTransactionBuilder {
     const algosdkTxn = decodedTxn.txn;
 
     if (decodedTxn.signed) {
-      this._transaction.signedTransaction =
-        typeof rawTransaction === 'string' ? Utils.hexStringToUInt8Array(rawTransaction) : rawTransaction;
+      this._transaction.signedTransaction = decodedTxn.rawTransaction;
+
```
Comment (rliberal-simtlix) [inline modules/account-lib/src/coin/algo/transactionBuilder.ts:296]: remove extra line

**Review by rliberal-simtlix — CHANGES_REQUESTED** [review]

**Review by rliberal-simtlix — APPROVED** [review]

**Review by Krin2 — APPROVED** [review]
LGTM

File: modules/account-lib/src/utils/crypto.ts (hunk 7 above)
Comment (Krin2) [reply modules/account-lib/src/utils/crypto.ts]: errased

File: modules/account-lib/src/coin/algo/utils.ts [hunk 14]
```diff
@@ -89,48 +123,82 @@ export class Utils implements BaseUtils {
...
+      sha512.sha512_256.array(decoded.seed).slice(SEED_BYTES_LENGTH - ALGORAND_CHECKSUM_BYTE_LENGTH, SEED_BYTES_LENGTH),
+    );
+
+    // Check if the checksum and the seed are equal
+    return _.isEqual(checksum, decoded.checksum);
   }
 
   /**
-   * Decodes a seed
+   * encodeSeed decodes an algo seed
```
Comment (CapnMigraine) [inline modules/account-lib/src/coin/algo/utils.ts:149]: this is not correct

File: modules/account-lib/src/coin/algo/utils.ts [hunk 15]
```diff
@@ -89,48 +123,82 @@ export class Utils implements BaseUtils {
...
+    const decoded = base32.decode.asBytes(seed);
+
+    // Sanity check
+    if (decoded.length !== ALGORAND_SEED_BYTE_LENGTH) throw new Error('seed seems to be malformed');
     return {
       seed: new Uint8Array(decoded.slice(0, ALGORAND_SEED_BYTE_LENGTH - ALGORAND_CHECKSUM_BYTE_LENGTH)),
-      checksum: new Uint8Array(decoded.slice(ALGORAND_NACL_SEED_BYTES_LENGTH, ALGORAND_SEED_BYTE_LENGTH)),
-    }
+      checksum: new Uint8Array(decoded.slice(SEED_BYTES_LENGTH, ALGORAND_SEED_BYTE_LENGTH)),
+    };
```
Comment (CapnMigraine) [inline modules/account-lib/src/coin/algo/utils.ts:190]: Missing all these methods: https://github.com/BitGo/BitGoJS/compare/BG-34606-algo-platform-accountlib-migration#diff-1362121543d1f6184f7fa744d680bc2764d55791c8a7aa7b3eb3143395ebd333R459-R552

do we need them?

File: modules/account-lib/src/coin/algo/ifaces.ts [hunk 16]
```diff
@@ -27,9 +27,17 @@ export interface Address {
   checksum: Uint8Array;
 }
```
[the remark this answers is not in the document]
Comment (Krin2) [reply modules/account-lib/src/coin/algo/ifaces.ts:29]: Some of these functionalities were already done so they don't appeared as changed, but they are in the code.
image

File: modules/account-lib/src/coin/algo/utils.ts [hunk 17]
```diff
@@ -89,48 +123,82 @@ export class Utils implements BaseUtils {
...
    *
-   * @param {string} seed - the seed to decode
-   * @returns {object} - a decoded seed that contains the seed and a checksum
+   * @param  {Buffer} secretKey - the secretKey valid.
+   * @returns {string} seed - the seed to be validated.
    */
-  protected decodeSeed(seed: string) {
-    if (typeof seed !== "string") throw ALGORAND_MALFORMED_SEED_ERROR; // try to decode
+  encodeSeed(secretKey: Buffer): string {
+    // return algosdk.encodeAddress(seed);
```
[the remark this answers is not in the document]
Comment (Krin2) [reply modules/account-lib/src/coin/algo/utils.ts:159]: yes, should be delete --> Done

File: modules/account-lib/src/coin/algo/txnSchema.ts [hunk 18]
```diff
@@ -62,3 +62,8 @@ export const AssetTransferTxnSchema = joi.object({
   assetAmount: joi.custom((val) => typeof val === 'number' || typeof val === 'bigint').required(),
   receiver: addressSchema.required(),
 });
+
```
[the remark this answers is not in the document]
Comment (Krin2) [reply modules/account-lib/src/coin/algo/txnSchema.ts:65]: These change is needed. It is in the code
image

File: modules/account-lib/src/coin/algo/utils.ts (hunk 14 above)
Comment (Krin2) [reply modules/account-lib/src/coin/algo/utils.ts:149]: done

File: modules/account-lib/src/coin/algo/utils.ts (hunk 15 above)
Comment (Krin2) [reply modules/account-lib/src/coin/algo/utils.ts:190]: Many of the methods that figured there were already implemented. In this PR were added just  isSeedValid, encodeSeed and decodeSeed.
https://github.com/BitGo/BitGoJS/commit/201085b74338fb3114752e41b3e4792668c39014#diff-1362121543d1f6184f7fa744d680bc2764d55791c8a7aa7b3eb3143395ebd333R394-R458

**Review by rliberal-simtlix — APPROVED** [review]

**Review by CapnMigraine — APPROVED** [review]
lgtm

**Pull request merged**